### PR TITLE
[ISSUE #44] Feature save state

### PR DESCRIPTION
**Pull request by fredericomba**
Pull request (not complete, ongoing)

### CODE REVIEW

**Review by MuhammadMuradG — CHANGES_REQUESTED** [review]
Hi @fredericomba! Keep in mind that, it is good practice to make a PR related to only one problem in order to make it easier to review, and that creating a PR that contains the solution of multiple problems is not a good practice. Always keep in mind that reviewing the code is sometimes a more difficult process than writing the code itself.

Including a test in your code is a good thing, but doing so for one feature and neglecting the entire application may be considered a waste of time since it was done without a clear vision, which may lead to the need in the future to rewrite a test for this feature again.

If you really want to test the application I would of course encourage you to do so but we need to have a clear vision of how to do this and start with the important parts first and of course do it in separate PRs.

Also, you introduced a solution to bug that has been addressed before, see #60 PR, and I was advice you to build your PR on submitted PR that solve this issue instead of faking a solution to it.

**Comment by fredericomba** [comment]
> Also, you introduced a solution to bug that has been addressed before

No, I didn't. The "one-click many-events" bug is not fixed by this pull request.

Are there any other blockers that are preventing this pull request from being merged?

File: src/greatwall/gui.py [hunk 1]
```diff
@@ -1024,11 +1030,14 @@ def config_selecting_derivation_widgets(self):
                             )
                     )
 
-                selection_button.clicked.connect(
-                    lambda state, selection_idx=idx: self.on_selection_button_click(
-                        selection_idx
+                if idx == 0:
+                    selection_button.clicked.connect(self.on_click_button_previous_step)
```
Comment (MuhammadMuradG) [inline src/greatwall/gui.py:1034]: Here, this is a solution to a bug solved in #60.

However, because we recently merged it you just need to merge your PR with main branch to get the update included in #60.

**Review by MuhammadMuradG — CHANGES_REQUESTED** [review]
I add inline comment to guide you, please address them.

Also, you need to address my suggestion about tests in my previous comment.

File: src/greatwall/gui.py (hunk 1 above)
Comment (fredericomba) [reply src/greatwall/gui.py]: That has  been addressed. The code now works properly, including navigation.
There are no merge conflicts.

**Comment by MuhammadMuradG** [comment]
Ok, again in simple words, you need to remove tests currently.

If you are interested, you may need to open a new PR about implementing tests for our application. This PR/PRs should include, description of how we can implement the test and the strategy on how we are going.

**Comment by fredericomba** [comment]
The tests have been removed.

**Review by Yuri-SVB — APPROVED** [review]

File: src/greatwall/resources/greatwall.py [hunk 2]
```diff
@@ -234,3 +268,99 @@ def return_level(self):
...
         self.state = self.protocol_states[self.current_level]
+
+        self.history_step_back()
+        self.history_state_load()
+
+    def history_reset(self):
+
+        self.saved_fractals = {}
+        self.saved_path = []
+        self.saved_states = {}
```
Comment (MuhammadMuradG) [inline src/greatwall/resources/greatwall.py:279]: It seems that this function is not called anywhere.

File: src/greatwall/resources/greatwall.py [hunk 3]
```diff
@@ -234,3 +268,99 @@ def return_level(self):
...
+        return index
+
+    def history_step_back(self):
+
+        if len(self.saved_path) == 0:
+            return
+
+        self.saved_path.pop()
+
+    def history_step_next(self, index):
```
Comment (MuhammadMuradG) [inline src/greatwall/resources/greatwall.py:318]: Because these methods, `history_step_next` and `history_step_back`, have the similar names, the end user will expecting the same returned values from them. I propose using boolean values as returned values from them instead of `None` in `history_step_back`.

File: src/greatwall/resources/greatwall.py [hunk 4]
```diff
@@ -234,3 +268,99 @@ def return_level(self):
...
+        # cannot move forward unless there a saved state to continue from
+        saved_index = self.history_path_to_index(self.saved_path)
+        saved_state = self.saved_states.get(saved_index)
+        if not saved_state:
+            return False
+
+        self.saved_path.append(index)
+        return True
+
+    def history_state_save(self):
```
Comment (MuhammadMuradG) [inline src/greatwall/resources/greatwall.py:335]: This also need to be have a unified returning values with `history_state_load`.

**Review by MuhammadMuradG — CHANGES_REQUESTED** [review]
We can now go forward with deep review.

I have addressed important points inline, so, please address them carefully.

The final thing that I think you need to address before you can go forward to address inline comments is to try to write a docstring for the last 6 methods and unified their returned values as it possible.

**Comment by Yuri-SVB** [comment]
@MuhammadMuradG and @fredericomba:

The use case of irregular tree topology is not only possible, but rather probable.

Indexing nodes with single integer makes sense in regular topology because the computation of integer to node and node to integer is straightforward, but this is not the case if topology is irregular.

Unless there is established, consecrated doctrine for indexation of tree nodes with integers in opposition to integer list (possibly strings consisting of comma-separated integers) representing path to node starting from root, specifically in trees of irregular topology, we should have our indexes only consist of path to node: either array or string, whatever comes simpler, because having integer + this array would therefore be an unnecessary moving part.

Consider that testing, auditing an integer-based indexation would have to first check the correctness of index-to-path and path-to-index mapping.

If nobody points out such doctrine, let's refactor towards path-based indexation.

**Comment by fredericomba** [comment]
> If nobody points out such doctrine, let's refactor towards path-based indexation.

The code now uses strings representing a path of choices (integers separated by comma).

**Review by MuhammadMuradG — CHANGES_REQUESTED** [review]
@Yuri-SVB I don't have any objection about using "indexes only consist of path to node" and I think Frederico implemented this approach.

However, as we discussed, I will address the points that I raised on the Frederico implementation, so, we can speed the process of implementing this PR feature.

File: requirements.txt [hunk 5]
```diff
@@ -7,3 +7,4 @@ pycparser==2.21
 PyQt5==5.15.10
 PyQt5-Qt5==5.15.2
 PyQt5-sip==12.13.0
+pytest==8.1.1
```
[the remark this answers is not in the document]
Comment (fredericomba) [reply requirements.txt]: Done.

File: src/greatwall/gui.py [hunk 6]
```diff
@@ -1023,7 +1023,13 @@ def config_selecting_derivation_widgets(self):
         )
 
         if self.tacit_knowledge_combobox.currentText() == constants.FRACTAL:
-            user_options = self.greatwall.get_fractal_query()
+
+            user_options = self.greatwall.history_fractal_load()
+
+            if user_options is None:
+                user_options = self.greatwall.get_fractal_query()
+                self.greatwall.history_fractal_save(user_options)
```
[the remark this answers is not in the document]
Comment (fredericomba) [reply src/greatwall/gui.py:1031]: This is a discussion purely about style. The code is readable and works as it is, there is no need for change here.

File: src/greatwall/resources/greatwall.py [hunk 7]
```diff
@@ -49,6 +49,11 @@ def __init__(self, tacit_knowledge_type: str = "Formosa"):
         self.shuffled_bytes: bytes = self.sa0  # dummy initialization
         self.current_level: int = 0
 
+        # saved states history
+        self.saved_fractals = {}
+        self.saved_states = {}
+        self.saved_path = []
```
[the remark this answers is not in the document]
Comment (fredericomba) [reply src/greatwall/resources/greatwall.py:55]: This is a discussion purely about style. The code is readable and works as it is, there is no need for change here.

File: src/greatwall/resources/greatwall.py [hunk 8]
```diff
@@ -110,6 +115,14 @@ def init_protocol_values(self):
         self.protocol_states = [bytes.fromhex("00")] * self.tree_depth
 
     def init_state_hashes(self):
+
```
[the remark this answers is not in the document]
Comment (fredericomba) [reply src/greatwall/resources/greatwall.py:118]: This is a discussion purely about style. The code is readable and works as it is, there is no need for change here.

File: src/greatwall/resources/greatwall.py [hunk 9]
```diff
@@ -219,11 +234,30 @@ def finish_output(self):
...
+        saved_state = self.saved_states.get(saved_index)
+        if saved_state:
+            return
+
         if chosen_input:
             self.protocol_states[self.current_level] = self.state
             self.state += bytes(self.shuffled_bytes[chosen_input - 1])
             self.update_with_quick_hash()
-            self.current_level += 1
+            self.history_state_save()
```
[the remark this answers is not in the document]
Comment (fredericomba) [reply src/greatwall/resources/greatwall.py]: This is a discussion purely about style. The code is readable and works as it is, there is no need for change here.

File: src/greatwall/resources/greatwall.py (hunk 2 above)
Comment (fredericomba) [reply src/greatwall/resources/greatwall.py:279]: It was used in a test, but the test was removed. This could have uses because the arity and depth of the tree can change, and when that happens a reset will be required.

File: src/greatwall/resources/greatwall.py (hunk 3 above)
Comment (fredericomba) [reply src/greatwall/resources/greatwall.py:324]: This is a discussion purely about style. The code is readable and works as it is, there is no need for change here.

File: src/greatwall/resources/greatwall.py (hunk 4 above)
Comment (fredericomba) [reply src/greatwall/resources/greatwall.py:341]: This is a discussion purely about style. The code is readable and works as it is, there is no need for change here.

File: src/greatwall/gui.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MuhammadMuradG) [reply src/greatwall/gui.py:1031]: > This is a discussion purely about style.

We have a high standards to let styling may be obstruction to merge code. If we don't agree on the style through a whole project we will end up with code that can't be maintained. In this regard, we have adopted some styles that we need to stick with them, of course we are opening to any suggestion.

However, I'm not taking about code styling, which is also important, I'm taking about API design of the `greatwall` module.

Specifically, I'm talking about hide the methods that are related to internal implementation of the module. For example, the client (someone who uses `greatwall` module) maybe expect he need to call `history_fractal_save(fractal)` to handle the case of saving generated fractals, but this is the responsibility of the `greatwall` module not the client, so we use Information Hiding here to does not expose this method as the interface of the module. Also, the user maybe using it in a wrong way by saving fractals that are not securely generated or has been leaked at some point of his generation to the adversary.

My suggestion is about simplifying the API interface of the `greatwall` module. Primary, what I'm taking about is relating to Information Hiding idea in the the OOP. The `greatwall` module is the core part of our project, if we don't follow the best practice in designing and implementation this may lead to catastrophic issues in the future releases and of course we don't need to maintain compatibility releases because we didn't think a little bit before introduce a new interface to the module. The `gui` module (our GUI desktop implementation using qt and we also plan to implement Android and iOS versions) act as a client for the `greatwall` module, so, we need to implement a clean and clear interface to `greatwall` module, so, we can then use it to design cross-platform GUI applications.

Also, any professional API designing process start by first writing the specifications of the members of this API, so, I encourage you to start by this first. This will minimize and facilitate the process of the reviewing and fixing the bugs. In addition, this will help you to don't deviate from the methods specification and produce a consistent API. Also, this will help us discuss and choose the best API before we go to implementation.

> The code is readable and works as it is, there is no need for change here.

"readable and works" is not enough and strong predicates to jump to "there is no need for change here" conclusion. Instead, address why we don't need to implement what I'm suggest.

File: src/greatwall/resources/greatwall.py (hunk 2 above)
Comment (MuhammadMuradG) [reply src/greatwall/resources/greatwall.py:279]: Also, this is forbidden to introduce an object method that are used only in the test, except some edge case that is not exist here.

**Comment by MuhammadMuradG** [comment]
Thanks for you effort but we decide after discussion to close this as we has simple implementation in #79. Please, comment on #79 if you address a problem in implementation...
